pkg/rtc/mediatrackreceiver.go

```diff
@@ -89,28 +100,83 @@ func NewMediaTrackReceiver(params MediaTrackReceiverParams) *MediaTrackReceiver
 
 func (t *MediaTrackReceiver) Restart() {
 	t.lock.Lock()
-	receiver := t.receiver
+	receivers := t.receivers
 	t.lock.Unlock()
 
-	if receiver != nil {
+	for _, receiver := range receivers {
 		receiver.SetMaxExpectedSpatialLayer(SpatialLayerForQuality(livekit.VideoQuality_HIGH))
-		t.MediaTrackSubscriptions.Restart()
 	}
+
+	t.MediaTrackSubscriptions.Restart()
 }
 
-func (t *MediaTrackReceiver) SetupReceiver(receiver sfu.TrackReceiver) {
+func (t *MediaTrackReceiver) SetupReceiver(receiver sfu.TrackReceiver, priority int, mid string) {
 	t.lock.Lock()
-	t.receiver = receiver
+	t.receivers = append(t.receivers, &simulcastReceiver{TrackReceiver: receiver, priority: priority})
+	sort.Slice(t.receivers, func(i, j int) bool {
+		return t.receivers[i].Priority() < t.receivers[j].Priority()
+	})
+
+	if mid != "" {
+		if priority == 0 {
+			t.params.TrackInfo.MimeType = receiver.Codec().MimeType
+			t.params.TrackInfo.Mid = mid
+
+			// for clients don't have simulcast codecs (old version or single codec), add the primary codec
+			if len(t.params.TrackInfo.Codecs) == 0 && t.Kind() == livekit.TrackType_VIDEO {
+				t.params.TrackInfo.Codecs = append(t.params.TrackInfo.Codecs, &livekit.SimulcastCodecInfo{})
+			}
+		}
+
+		for i, ci := range t.params.TrackInfo.Codecs {
+			if i == priority {
+				ci.Mid = mid
+				ci.MimeType = receiver.Codec().MimeType
+			}
+		}
+	}
+
 	t.lock.Unlock()
+	t.params.Logger.Debugw("setup receiver", "mime", receiver.Codec().MimeType, "priority", priority, "receivers", t.receivers)
+	t.MediaTrackSubscriptions.AddCodec(receiver.Codec().MimeType)
 
 	t.MediaTrackSubscriptions.Start()
 }
 
-func (t *MediaTrackReceiver) ClearReceiver() {
+func (t *MediaTrackReceiver) SetLayerSsrc(mime string, rid string, ssrc uint32) {
 	t.lock.Lock()
-	t.receiver = nil
+	defer t.lock.Unlock()
+
+	layer := sfu.RidToLayer(rid)
+	for _, receiver := range t.receivers {
+		if strings.EqualFold(receiver.Codec().MimeType, mime) && int(layer) < len(receiver.layerSSRCs) {
+			receiver.layerSSRCs[layer] = ssrc
+			return
+		}
+	}
+}
+
+func (t *MediaTrackReceiver) ClearReceiver(mime string) {
+	t.lock.Lock()
+	for idx, receiver := range t.receivers {
+		if strings.EqualFold(receiver.Codec().MimeType, mime) {
+			t.receivers[idx] = t.receivers[len(t.receivers)-1]
+			t.receivers = t.receivers[:len(t.receivers)-1]
+			break
+		}
+	}
+	closeSubscription := len(t.receivers) == 0
```

Comment:
We probably need a shadow copy just to be safe Jie. There is another place where a shadow copy is ranged over.

Reply:
will do that later in a separate PR